server/encode/encode.go

```diff
@@ -36,10 +39,11 @@ func fileExists(filename string) bool {
 
 func Profiles() map[string]Profile {
 	return map[string]Profile{
-		"mp3":     {"mp3", 128, []string{"-c:a", "libmp3lame"}, false},
-		"mp3_rg":  {"mp3", 128, []string{"-c:a", "libmp3lame"}, true},
-		"opus":    {"opus", 96, []string{"-c:a", "libopus", "-vbr", "constrained"}, false},
-		"opus_rg": {"opus", 96, []string{"-c:a", "libopus", "-vbr", "constrained"}, true},
+		"mp3":      {"mp3", 128, []string{"-c:a", "libmp3lame"}, false, false, false},
+		"mp3_rg":   {"mp3", 128, []string{"-c:a", "libmp3lame"}, true, false, false},
+		"opus":     {"opus", 96, []string{"-c:a", "libopus", "-vbr", "on"}, false, false, false},
+		"opus_rg":  {"opus", 96, []string{"-c:a", "libopus", "-vbr", "on"}, true, false, false},
+		"opus_car": {"opus", 96, []string{"-c:a", "libopus", "-vbr", "on"}, true, true, true},
```

Comment:
now with all these booleans, it's little bit hard to know what's going on with for example `false, false, false`

maybe we could add some field names, and use go's false bool zero-value?

```suggestion
		"mp3": {
			Format:        "mp3",
			Bitrate:       128,
			ffmpegOptions: []string{"-c:a", "libmp3lame"},
		},
		"mp3_rg": {
			Format:        "mp3",
			Bitrate:       128,
			ffmpegOptions: []string{"-c:a", "libmp3lame"},
			forceRG:       true,
		},
		"opus": {
			Format:        "opus",
			Bitrate:       96,
			ffmpegOptions: []string{"-c:a", "libopus", "-vbr", "on"},
		},
		"opus_rg": {
			Format:        "opus",
			Bitrate:       96,
			ffmpegOptions: []string{"-c:a", "libopus", "-vbr", "on"},
			forceRG:       true,
		},
		"opus_car": {
			Format:        "opus",
			Bitrate:       96,
			ffmpegOptions: []string{"-c:a", "libopus", "-vbr", "on"},
			forceRG:       true,
			hiGainRG:      true,
			upsample:      true,
		},
```

Reply:
I was thinking about converting `forceRG` to an Enum (or something similar) but never tried it.
Overall this format seems nice, but it's starting to look like it's only a matter of time before we have to move it "outside" to a YAML config file or some other place. :D

Reply:
yeah the enum is interesting. do you  think something like this could work?

```diff
diff --git a/server/encode/encode.go b/server/encode/encode.go
index 550b6f7..58d0a0b 100644
--- a/server/encode/encode.go
+++ b/server/encode/encode.go
@@ -20,13 +20,20 @@ const (
 	ffmpeg  = "ffmpeg"
 )
 
+type ReplayGain int
+
+const (
+	RGNone ReplayGain = iota
+	RGForce
+	RGHigh
+)
+
 type Profile struct {
-	Format        string
-	Bitrate       int
-	ffmpegOptions []string
-	forceRG       bool
-	hiGainRG      bool
-	upsample      bool
+	Format     string
+	Bitrate    int
+	Options    []string
+	ReplayGain ReplayGain
+	Upsample   bool
 }
 
 func fileExists(filename string) bool {
@@ -39,11 +46,35 @@ func fileExists(filename string) bool {
 
 func Profiles() map[string]Profile {
 	return map[string]Profile{
-		"mp3":      {"mp3", 128, []string{"-c:a", "libmp3lame"}, false, false, false},
-		"mp3_rg":   {"mp3", 128, []string{"-c:a", "libmp3lame"}, true, false, false},
-		"opus":     {"opus", 96, []string{"-c:a", "libopus", "-vbr", "on"}, false, false, false},
-		"opus_rg":  {"opus", 96, []string{"-c:a", "libopus", "-vbr", "on"}, true, false, false},
-		"opus_car": {"opus", 96, []string{"-c:a", "libopus", "-vbr", "on"}, true, true, true},
+		"mp3": {
+			Format:  "mp3",
+			Bitrate: 128,
+			Options: []string{"-c:a", "libmp3lame"},
+		},
+		"mp3_rg": {
+			Format:     "mp3",
+			Bitrate:    128,
+			Options:    []string{"-c:a", "libmp3lame"},
+			ReplayGain: RGForce,
+		},
+		"opus": {
+			Format:  "opus",
+			Bitrate: 96,
+			Options: []string{"-c:a", "libopus", "-vbr", "on"},
+		},
+		"opus_rg": {
+			Format:     "opus",
+			Bitrate:    96,
+			Options:    []string{"-c:a", "libopus", "-vbr", "on"},
+			ReplayGain: RGForce,
+		},
+		"opus_car": {
+			Format:     "opus",
+			Bitrate:    96,
+			Options:    []string{"-c:a", "libopus", "-vbr", "on"},
+			ReplayGain: RGHigh,
+			Upsample:   true,
+		},
 	}
 }
 
@@ -96,35 +127,34 @@ func ffmpegCommand(filePath string, profile Profile) (*exec.Cmd, error) {
 		"-vn",
 		"-b:a", fmt.Sprintf("%dk", profile.Bitrate),
 	}
-	args = append(args, profile.ffmpegOptions...)
-	if profile.forceRG {
-		aBaselineGain := 6
-		if profile.hiGainRG {
-			// This baseline gain results in final track being +3~5dB louder
-			// than Foobar2000's default ReplayGain target volume.
-			// This makes it easier to listen to music in a car, where all other
-			// sources are usually ten thousand times louder than RG-adjusted music.
-			// -- @spijet
-			aBaselineGain = 15
-		}
-		aFilters := []string{
-			fmt.Sprintf("volume=replaygain=track:replaygain_preamp=%ddB:replaygain_noclip=0", aBaselineGain),
-			"alimiter=level=disabled",
-			"asidedata=mode=delete:type=REPLAYGAIN",
-		}
+	args = append(args, profile.Options...)
 
-		// opus always forces output to 48kHz sampling rate, but we can still use upsampling
-		// to increase RG and alimiter's peak limiting precision, which is desirable in some
-		// cases. ffmpeg's `soxr` resampler is quite fast on x86-64: it takes around 5 seconds
-		// on my Ryzen 3600 to transcode an 8-minute FLAC with 2x upsample and RG applied.
+	var aFilters []string
+	// opus always forces output to 48kHz sampling rate, but we can still use upsampling
+	// to increase RG and alimiter's peak limiting precision, which is desirable in some
+	// cases. ffmpeg's `soxr` resampler is quite fast on x86-64: it takes around 5 seconds
+	// on my Ryzen 3600 to transcode an 8-minute FLAC with 2x upsample and RG applied.
+	// -- @spijet
+	if profile.Upsample {
+		aFilters = append(aFilters, "aresample=96000:resampler=soxr")
+	}
+
+	switch profile.ReplayGain {
+	case RGForce:
+		aFilters = append(aFilters, ffmpegPreamp(6)...)
+	case RGHigh:
+		// This baseline gain results in final track being +3~5dB louder
+		// than Foobar2000's default ReplayGain target volume.
+		// This makes it easier to listen to music in a car, where all other
+		// sources are usually ten thousand times louder than RG-adjusted music.
 		// -- @spijet
-		if profile.upsample {
-			aFilters = append([]string{"aresample=96000:resampler=soxr"}, aFilters...)
-		}
-		aFilterString := strings.Join(aFilters, ", ")
+		aFilters = append(aFilters, ffmpegPreamp(15)...)
+	}
+
+	if len(aFilters) > 0 {
 		args = append(args,
 			// set up replaygain processing
-			"-af", aFilterString,
+			"-af", strings.Join(aFilters, ", "),
 			// drop redundant replaygain tags
 			"-metadata", "replaygain_album_gain=",
 			"-metadata", "replaygain_album_peak=",
@@ -134,6 +164,7 @@ func ffmpegCommand(filePath string, profile Profile) (*exec.Cmd, error) {
 			"-metadata", "r128_track_gain=",
 		)
 	}
+
 	args = append(args, "-f", profile.Format, "-")
 	ffmpegPath, err := exec.LookPath(ffmpeg)
 	if err != nil {
@@ -144,6 +175,14 @@ func ffmpegCommand(filePath string, profile Profile) (*exec.Cmd, error) {
 	// but please do let me know if you see otherwise
 }
 
+func ffmpegPreamp(dB int) []string {
+	return []string{
+		fmt.Sprintf("volume=replaygain=track:replaygain_preamp=%ddB:replaygain_noclip=0", dB),
+		"alimiter=level=disabled",
+		"asidedata=mode=delete:type=REPLAYGAIN",
+	}
+}
+
 func encode(out io.Writer, trackPath, cachePath string, profile Profile) error {
 	// prepare cache part file path
 	cachePartPath := fmt.Sprintf("%s.part", cachePath)
```
![image](https://user-images.githubusercontent.com/6832539/122983611-1839e500-d394-11eb-96bd-fe93d1e8b4d4.png)

Reply:
Yep, looking good!
As for `len(aFilters) > 0` check — it seems sufficient for now, because we only use audio filters when we're forcing ReplayGain, but it may lead to `ffmpeg` stripping out RG tags if/when we add some encoding profile with other filters in it (say, a profile that does on-the-fly stereo⇒binaural conversion for headphone listening).

What if we set it up like this?
```golang
var stripRGTags bool
switch profile.ReplayGain {
case RGForce:
    stripRGTags = true
...
}

if len(aFilters) > 0 {
    args = append(args, "-af", strings.Join(aFilters, ", "))
    if stripRGTags {
        args = append(args, "-metadata", ...)
    }
}
```

Reply:
ah good shout, thanks!
then in that case, as we want to add more functionality to the ffmpegCommand function, it might get a bit complicated in future.
what about a "functional options" pattern - where we can centralise the the different behaviours of each profile - and possibly make adding future functionality easier?

it could look something like this
![image](https://user-images.githubusercontent.com/6832539/123009120-0c5e1b00-d3b4-11eb-9b23-9b5c08d1e982.png)


https://termbin.com/p7ee

(not at all tested, but just wondering if you like this idea)

Reply:
I was thinking about it too, but then I thought about a reverse approach, similar to what they do in Subsonic/MadSonic/AirSonic, where all encoding profiles are just four-field maps (converted to YAML for readability):
```yaml
# Simple example, FLAC → MP3
- sourceFormat: flac
  outputFormat: mp3
  # Format tokens: %s is source file, %f is output format, %b is bitrate
  stageOne: "ffmpeg -i %s -f %f -b:a %bk pipe:1"
  stageTwo:
# Two-stage example, some tracker module format → MP3
- sourceFormat: s3m
  outputFormat: mp3
  stageOne: "xmp --nocmd --stdout --bits 16 --frequency 44100 %s"
  stageTwo: "ffmpeg -i pipe:0 -f %f -b:a %bk pipe:1"
```

These allow for much more flexibility, but IMO are harder for end-users to make sense of. What do you think?

Reply:
I would be in favour of this :+1: seems nice a simple

though if it were a real yaml file, I'd be hesitant to add a second configuration file to gonic. 
maybe an idea could be to store the profiles as structs first? then later to make them configurable my the web UI. this would help with #121

Reply:
You mean storing them in gonic's database? That sounds good, yeah. As for configuration files — a YAML file could be used as a "preseed", so the users could feed their profiles into gonic during the first deploy when using Docker.